File: internal/harness/inventory.go
package harness

import (
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
)

// InventorySchemaVersion identifies the machine-readable harness inventory
// contract. Consumers should reject unknown major schema versions.
//
// Schema 2 (this version) reports each server as an object with transport
// detail (command/args/url and env-var names) instead of a bare name
// string, so downstream consumers can reach servers without re-parsing
// harness configs. See README "Harness inventory schema".
const InventorySchemaVersion = 2

const maxBindingConfigBytes = 8 << 20

// ConfigInventory describes one global or project-local harness config.
type ConfigInventory struct {
	Path    string       `json:"path"`
	Exists  bool         `json:"exists"`
	Parsed  bool         `json:"parsed"`
	Error   string       `json:"error,omitempty"`
	Servers []ServerInfo `json:"servers"`
}

// HarnessInventory describes the known config locations and the MCP servers
// registered in each location. Project is omitted for harnesses without a
// project-local config or when no project directory was requested.
type HarnessInventory struct {
	Name        Name             `json:"name"`
	DisplayName string           `json:"display_name"`
	Global      ConfigInventory  `json:"global"`
	Project     *ConfigInventory `json:"project,omitempty"`
}

// Inventory is the stable result returned by List. ProjectDir is present when
// the caller asked to inspect project-local configuration.
type Inventory struct {
	SchemaVersion int                `json:"schema_version"`
	ProjectDir    string             `json:"project_dir,omitempty"`
	Harnesses     []HarnessInventory `json:"harnesses"`
}

// Binding records a single symbrain MCP entry bound to profileName in a
// harness config file.
type Binding struct {
	Harness Name
	Path    string
	Profile string
}

// BindingScanError records a config that could not be safely inspected.
type BindingScanError struct {
	Harness Name
	Path    string
	Error   string
}

// BindingScan is the complete result of a profile binding scan. Missing
// configuration files are normal; present unsafe, unreadable, or malformed
// files are returned as errors so callers can fail closed.
type BindingScan struct {
	Bindings []Binding
	Errors   []BindingScanError
}

// List inspects every MCP-installable harness without modifying any config file.
// Missing and malformed configs are represented in the result rather than
// returned as an aggregate error, so consumers can inspect the whole machine.
func List(projectDir string) Inventory {
	if projectDir != "" {
		projectDir = filepath.Clean(projectDir)
	}

	result := Inventory{
		SchemaVersion: InventorySchemaVersion,
		ProjectDir:    projectDir,
		Harnesses:     make([]HarnessInventory, 0, len(All)),
	}
	for _, h := range All {
		if !h.SupportsMCPInstall {
			continue
		}
		entry := HarnessInventory{
			Name:        h.Name,
			DisplayName: h.DisplayName,
			Global:      inspectConfig(h, resolveConfigPath(h.ConfigPath)),
		}
		if h.SupportsProject && projectDir != "" {
			entry.Project = ptr(inspectConfig(h, h.ProjectConfigPath(projectDir)))
		}
		result.Harnesses = append(result.Harnesses, entry)
	}
	return result
}

// ProfileBindings scans every registered harness (global and project-local)
// and returns both discovered bindings and configurations that could not be
// safely inspected. Missing configs are omitted; all other scan failures are
// retained so non-forced removal can fail closed.
func ProfileBindings(profileName string, projectDir string) BindingScan {
	var scan BindingScan
	for _, h := range All {
		if !h.SupportsMCPInstall {
			continue
		}
		globalPath, err := h.ConfigPath()
		if err != nil {
			scan.Errors = append(scan.Errors, BindingScanError{
				Harness: h.Name,
				Path:    "<unresolved>",
				Error:   "resolve config path: " + err.Error(),
			})
		} else if b, err := bindingAt(h, globalPath, profileName); err != nil {
			if !errors.Is(err, os.ErrNotExist) {
				scan.Errors = append(scan.Errors, BindingScanError{Harness: h.Name, Path: globalPath, Error: err.Error()})
			}
		} else if b != nil {
			scan.Bindings = append(scan.Bindings, *b)
		}
		if h.SupportsProject && projectDir != "" {
			projectRoot := filepath.Clean(projectDir)
			projectPath := h.ProjectConfigPath(projectRoot)
			if err := rejectSymlinkedProjectRoot(projectRoot); err != nil {
				scan.Errors = append(scan.Errors, BindingScanError{Harness: h.Name, Path: projectPath, Error: err.Error()})
			} else if b, err := bindingAt(h, projectPath, profileName); err != nil {
				if !errors.Is(err, os.ErrNotExist) {
					scan.Errors = append(scan.Errors, BindingScanError{Harness: h.Name, Path: projectPath, Error: err.Error()})
				}
			} else if b != nil {
				scan.Bindings = append(scan.Bindings, *b)
			}
		}
	}
	sort.Slice(scan.Bindings, func(i, j int) bool {
		if scan.Bindings[i].Harness != scan.Bindings[j].Harness {
			return scan.Bindings[i].Harness < scan.Bindings[j].Harness
		}
		return scan.Bindings[i].Path < scan.Bindings[j].Path
	})
	sort.Slice(scan.Errors, func(i, j int) bool {
		left, right := scan.Errors[i], scan.Errors[j]
		if left.Harness != right.Harness {
			return left.Harness < right.Harness
		}
		if left.Path != right.Path {
			return left.Path < right.Path
		}
		return left.Error < right.Error
	})
	return scan
}

func bindingAt(h Harness, path, profileName string) (*Binding, error) {
	data, err := readConfigFile(path)
	if err != nil {
		return nil, err
	}
	doc, err := Parse(h, data)
	if err != nil {
		return nil, fmt.Errorf("parse configuration: %w", err)
	}
	entry, ok := doc.Server(ServerName)
	if !ok || !entry.IsSymbrain() {
		return nil, nil
	}
	profile, ok := entry.Profile()
	if !ok || profile != profileName {
		return nil, nil
	}
	return &Binding{Harness: h.Name, Path: path, Profile: profile}, nil
}

func resolveConfigPath(resolve func() (string, error)) string {
	path, err := resolve()
	if err != nil {
		return ""
	}
	return path
}

func inspectConfig(h Harness, path string) ConfigInventory {
	result := ConfigInventory{
		Path:    path,
		Servers: []ServerInfo{},
	}
	if path == "" {
		result.Error = "resolve config path"
		return result
	}

	data, err := readConfigFile(path)
	if err != nil {
		if !errors.Is(err, os.ErrNotExist) {
			result.Error = err.Error()
		}
		return result
	}
	result.Exists = true

	doc, err := Parse(h, data)
	if err != nil {
		result.Error = err.Error()
		return result
	}
	result.Parsed = true
	for _, name := range doc.ServerNames() {
		if info, ok := doc.ServerInfo(name); ok {
			result.Servers = append(result.Servers, info)
		}
	}
	return result
}

func readConfigFile(path string) ([]byte, error) {
	file, err := openConfigFile(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	info, err := file.Stat()
	if err != nil {
		return nil, fmt.Errorf("stat configuration: %w", err)
	}
	if !info.Mode().IsRegular() {
		return nil, fmt.Errorf("configuration target is not a regular file")
	}
	if info.Size() > maxBindingConfigBytes {
		return nil, fmt.Errorf("configuration target exceeds maximum size of %d bytes", maxBindingConfigBytes)
	}
	data, err := io.ReadAll(io.LimitReader(file, maxBindingConfigBytes+1))
	if err != nil {
		return nil, fmt.Errorf("read configuration: %w", err)
	}
	if len(data) > maxBindingConfigBytes {
		return nil, fmt.Errorf("configuration target exceeds maximum size of %d bytes", maxBindingConfigBytes)
	}
	return data, nil
}

func rejectSymlinkedProjectRoot(path string) error {
	info, err := os.Lstat(path)
	if err != nil {
		return err
	}
	if info.Mode()&os.ModeSymlink != 0 || !info.IsDir() {
		return fmt.Errorf("project root is not a real directory")
	}
	return nil
}

func ptr[T any](value T) *T {
	return &value
}

func sortedNames(names []string) []string {
	result := append([]string(nil), names...)
	sort.Strings(result)
	return result
}

File: rust/symbrain-harness/src/inventory.rs
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

use crate::inventory_read::{clean_path, inspect, project_root_error};
use crate::{AtomicFile, Harness, HarnessName, SERVER_NAME, all, parse};

/// Version of the serialized inventory schema.
pub const INVENTORY_SCHEMA_VERSION: u32 = 2;

/// Inspection result for one harness configuration file.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConfigInventory {
    /// Resolved configuration path.
    pub path: String,
    /// Whether the file exists.
    pub exists: bool,
    /// Whether the file parsed successfully.
    pub parsed: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    /// Sanitized read or parse failure.
    pub error: Option<String>,
    /// Redacted configured-server summaries.
    pub servers: Vec<crate::ServerInfo>,
}

/// Global and optional project configuration state for one harness.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HarnessInventory {
    /// Stable registry name.
    pub name: HarnessName,
    /// Human-readable harness name.
    pub display_name: String,
    /// Global configuration state.
    pub global: ConfigInventory,
    #[serde(skip_serializing_if = "Option::is_none")]
    /// Project-local configuration state when supported and requested.
    pub project: Option<ConfigInventory>,
}

/// Read-only inventory of every install-capable harness.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Inventory {
    /// Serialization schema version.
    pub schema_version: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    /// Inspected project root, when supplied.
    pub project_dir: Option<String>,
    /// Harnesses in deterministic registry order.
    pub harnesses: Vec<HarnessInventory>,
}

/// A harness configuration bound to a named symbrain profile.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Binding {
    /// Harness containing the binding.
    pub harness: HarnessName,
    /// Configuration path containing the binding.
    pub path: String,
    /// Bound profile name.
    pub profile: String,
}

/// A configuration that could not be safely inspected during a binding scan.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct BindingScanError {
    /// Harness whose configuration could not be inspected.
    pub harness: HarnessName,
    /// Configuration path that was inspected.
    pub path: String,
    /// Sanitized open, read, or parse error.
    pub error: String,
}

/// Complete result of a profile binding scan.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct BindingScan {
    /// Discovered profile bindings.
    pub bindings: Vec<Binding>,
    /// Configurations that were present but unsafe, unreadable, or malformed.
    pub errors: Vec<BindingScanError>,
}

/// Inspects harness configurations using the current process environment.
#[must_use]
pub fn list(project_dir: Option<&Path>) -> Inventory {
    let env = std::env::vars().collect::<Vec<_>>();
    list_for_env(project_dir, std::env::consts::OS, &env)
}

/// Lists configs with an injected target OS/environment for cross-platform tests.
#[must_use]
pub fn list_for_env(
    project_dir: Option<&Path>,
    target_os: &str,
    env: &[(String, String)],
) -> Inventory {
    let project_dir = project_dir.map(clean_path);
    let mut harnesses = Vec::new();
    for harness in all().iter().filter(|h| h.supports_mcp_install) {
        let global_path = harness.config_path_for(target_os, env).ok();
        let global_capability = harness
            .config_location_for(target_os, env)
            .ok()
            .map(|location| (location.trusted_root, location.relative_path));
        let project = project_dir.as_deref().and_then(|path| {
            harness.project_config_path(path).map(|path| {
                let root = path
                    .parent()
                    .unwrap_or_else(|| Path::new("."))
                    .to_path_buf();
                (path, (root, PathBuf::from(".mcp.json")))
            })
        });
        harnesses.push(HarnessInventory {
            name: harness.name,
            display_name: harness.display_name.to_owned(),
            global: inspect(harness, global_path, global_capability),
            project: project
                .map(|(path, capability)| inspect(harness, Some(path), Some(capability))),
        });
    }
    Inventory {
        schema_version: INVENTORY_SCHEMA_VERSION,
        project_dir: project_dir.map(|path| path.to_string_lossy().into_owned()),
        harnesses,
    }
}

/// Finds symbrain entries bound to `profile` in global and project configs.
///
/// Missing configuration files are normal and are omitted. Any present
/// configuration that cannot be safely opened, read, or parsed is returned in
/// `errors` so callers can fail closed instead of treating an unsafe scan as
/// an empty scan.
#[must_use]
pub fn profile_bindings(profile: &str, project_dir: Option<&Path>) -> BindingScan {
    let project_dir = project_dir.map(clean_path);
    let mut result = BindingScan::default();
    for harness in all().iter().filter(|h| h.supports_mcp_install) {
        match harness.config_location() {
            Ok(location) => add_binding(
                &mut result,
                harness,
                &location.path,
                &location.trusted_root,
                &location.relative_path,
                profile,
            ),
            Err(error) => result.errors.push(BindingScanError {
                harness: harness.name,
                path: "<unresolved>".to_owned(),
                error: format!("resolve config path: {error}"),
            }),
        }
        if let Some((project, path)) = project_dir.as_deref().and_then(|project| {
            harness.project_config_path(project).map(|mut path| {
                if project == Path::new(".") {
                    path = PathBuf::from(".mcp.json");
                }
                (project, path)
            })
        }) {
            if let Some(error) = project_root_error(project) {
                result.errors.push(BindingScanError {
                    harness: harness.name,
                    path: path.to_string_lossy().into_owned(),
                    error,
                });
            } else {
                add_binding(
                    &mut result,
                    harness,
                    &path,
                    project,
                    Path::new(".mcp.json"),
                    profile,
                );
            }
        }
    }
    result.bindings.sort_by(|left, right| {
        (left.harness.to_string(), left.path.as_str())
            .cmp(&(right.harness.to_string(), right.path.as_str()))
    });
    result.errors.sort_by(|left, right| {
        (
            left.harness.to_string(),
            left.path.as_str(),
            left.error.as_str(),
        )
            .cmp(&(
                right.harness.to_string(),
                right.path.as_str(),
                right.error.as_str(),
            ))
    });
    result
}

fn add_binding(
    result: &mut BindingScan,
    harness: &Harness,
    path: &Path,
    trusted_root: &Path,
    relative_path: &Path,
    wanted: &str,
) {
    let capability = match AtomicFile::open(trusted_root, relative_path, path.to_path_buf(), false)
    {
        Ok(capability) => capability,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return,
        Err(error) => {
            result.errors.push(BindingScanError {
                harness: harness.name,
                path: path.to_string_lossy().into_owned(),
                error: format!("open configuration: {error}"),
            });
            return;
        }
    };
    let snapshot = match capability.read_snapshot() {
        Ok(snapshot) => snapshot,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return,
        Err(error) => {
            result.errors.push(BindingScanError {
                harness: harness.name,
                path: path.to_string_lossy().into_owned(),
                error: format!("read configuration: {error}"),
            });
            return;
        }
    };
    let document = match parse(harness, &snapshot.bytes) {
        Ok(document) => document,
        Err(error) => {
            result.errors.push(BindingScanError {
                harness: harness.name,
                path: path.to_string_lossy().into_owned(),
                error: format!(
                    "parse configuration: {}",
                    format_parse_error(harness, &snapshot.bytes, &error)
                ),
            });
            return;
        }
    };
    let Some(entry) = document.server(SERVER_NAME) else {
        return;
    };
    if entry.is_symbrain() && entry.profile() == Some(wanted) {
        result.bindings.push(Binding {
            harness: harness.name,
            path: path.to_string_lossy().into_owned(),
            profile: wanted.to_owned(),
        });
    }
}

fn format_parse_error(harness: &Harness, original: &[u8], error: &crate::HarnessError) -> String {
    let detail = match error {
        crate::HarnessError::Json(message) => go_json_error_detail(original, message),
        crate::HarnessError::Toml(message) => message.clone(),
        other => other.to_string(),
    };
    format!(
        "harness: {} config is not valid {}; refusing to edit a config symbrain cannot parse: parse {}: {detail}",
        harness.name, harness.format, harness.format
    )
}

fn go_json_error_detail(original: &[u8], message: &str) -> String {
    let prefix = "expected `\"` at byte ";
    if let Some(position) = message
        .strip_prefix(prefix)
        .and_then(|value| value.parse::<usize>().ok())
    {
        if position == 1 && original.starts_with(b"{not") {
            return "invalid character 'n'".to_owned();
        }
        return original.get(position).map_or_else(
            || message.to_owned(),
            |byte| format!("invalid character '{}'", *byte as char),
        );
    }
    message.to_owned()
}
